[Chua Zen Khoon] iP

ChuaZenKhoon opened the pull request:
# Balom

> "*Procrastination is something best put off till tomorrow.*" - Gerald Vaughan [(source)](https://www.askideas.com/63-best-procrastination-quotes-and-sayings/procrastination-is-something-best-put-off-until-tomorrow-gerald-vaughan/)

Balom is a type of chatbot that helps you with managing tasks that you have to do!

## Pros
* Easy to learn 👍 
* Neat and Tidy
* text-based
* Now has GUI for easy use!

## Stuff you can do
1. Add tasks and delete them
2. **Find** tasks
3. Mark and Unmark tasks
4. (NEW) Snooze tasks

## Visual Representation
- [x] \(Deadline) Read book by Oct 10 2020 18:00 

### Changes are automatically saved to your hard disk with `fileUpdate` and scanned when you reopen it!
Here is a snapshot of the chatbot working!



```java
public static void main(String[] args) {
        Duke balom = new Duke("C:/repos/cs2103t stuff/data", "duke.txt");
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ChuaZenKhoon committed:
Add basic skeleton

ChuaZenKhoon committed:
Add echo

ChuaZenKhoon committed:
Revert "Add echo"

This reverts commit 1af93215c597b59b2b60e6d418ed6450a1401133.

ChuaZenKhoon committed:
Reapply "Add echo"

This reverts commit a13b58677c129f26002bda8707ee303917705f8f.

ChuaZenKhoon committed:
Add add, list

User is now able to add input into a list and view it

ChuaZenKhoon committed:
Add Task class

User is now able to mark and unmark tasks in the list of inputs.
Chatbot now holds a Task array instead of a string array previously.
Code ensures invalid input of task.no and marking/unmarking already marked/unmarked tasks is handled.
Javadocs is done.

ChuaZenKhoon committed:
Add Todo, Event, Deadline

Include inheritance to have three types of taks user can handle
javadocs for new classes
abstracted out some functions to make main loop more readable

ChuaZenKhoon committed:
Add automated UI testing

Use of various input to test correct functionality of chatbot as of current: Level-4

ChuaZenKhoon committed:
Add Exception Handling

Handled: Invalid input, empty space, incomplete input, invalid marking/unmarking, remarking/unmarking, empty task list

ChuaZenKhoon committed:
Add support for deleting tasks from the list

User can now delete a task of valid index in the list
Automated UI testing is updated for current build

ChuaZenKhoon committed:
Minor Adjustments for Level-6

Add javadocs, shift arraylist into constructor for chatbot, update automated UI testing input

ChuaZenKhoon committed:
Minor changes

ChuaZenKhoon committed:
Add Save to hard disk

ChuaZenKhoon committed:
Add Date Time Management

Chatbot is now able to recognise Date Time for
Event/Deadline

ChuaZenKhoon committed:
Add more OOP

Separated faculties of the chatbot into relevant Classes

ChuaZenKhoon committed:
Add packages

Shifted all classes into duke package

ChuaZenKhoon committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ChuaZenKhoon committed:
Add A-gradle

Able to run Duke with gradle

ChuaZenKhoon committed:
Add JUnit

Configured test cases for 2 methods in
2 classes, minor adjustments for exceptions

ChuaZenKhoon committed:
Add Javadocs

ChuaZenKhoon committed:
Add CodingStandard

ChuaZenKhoon committed:
Minor changes

ChuaZenKhoon committed:
Add Find

Allows user to find tasks

ChuaZenKhoon committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Task.java

ChuaZenKhoon committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Ui.java

jskimdev commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,131 @@
+ * AY23/24 Semester 2
+ * Author: Chua Zen Khoon
+ */
+public class Duke {
```
Maybe change the name of Duke class to your chatbot name?

jskimdev commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,95 @@
+        String[] results = new String[3];
+
+        String keyword = echo.split(" ")[0];
+        if (keyword.equals("deadline")) {
```
maybe can create another boolean variable to increase readability?

isDeadine = keyword.equals("deadline)
...
if (isDeadline)

jskimdev commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,161 @@
+     * @throws DukeException when data file does not exist
+     */
+
+    public ArrayList<Task> dirAndFileSetUp() throws DukeException {
```
verb should come first for method name

jskimdev commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,161 @@
+     * @param  command whether the Task is to be updated/deleted from the file
+     * @param  lineNum at which line in the file the data is to be deleted
+     */
+    public void fileUpdate(Task taskToUpdate, int command, int lineNum) {
```
verb should come first in method naming

jskimdev commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,80 @@
+
+        if (echo.equals("bye") || echo.equals("Bye")) {
+            results[0] = 1;
+        } else if (echo.equals("list") || echo.equals("List")) {
```
can use echo.toLowerCase() method to convert all the letters to lowercase

ChuaZenKhoon replied:
Thanks for the suggestions and review!

jskimdev reviewed (commented):
Overall, it looks good to me. It is well-documented and easily readable code. I think it meets the most of coding standards and quality. Keep up the good work!

ChuaZenKhoon committed:
Add Checkstyle

Corrected style problems in all files

ChuaZenKhoon committed:
Act on peer review

implemented coding naming changes

ChuaZenKhoon committed:
Add GUI

Adjusted functions that printed to console
to now return string to be passed to
DialogBox

ChuaZenKhoon committed:
Update README.md

ChuaZenKhoon committed:
Fix truncated output

Current output in GUI is truncated
at a certain length, fixed with
solution on forum

ChuaZenKhoon committed:
Merge branch 'master' of https://github.com/ChuaZenKhoon/ip

Ragnapop commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,140 @@
+                output = tasks.showTasks();
+                break;
+            case 3:
+                output = tasks.markMechanism(result[1]);
+                break;
+            case 4:
+                output = tasks.unmarkMechanism(result[1]);
+                break;
+            case 5:
+                output = tasks.deleteMechanism(result[1]);
+                if (result[1] <= tasks.getSize() + 1) {
+                    storage.updateFile(new Task("To Delete"), 1, result[1]);
+                }
+                break;
+            case 6:
+                String keywordToFind = ui.analyseFind(echo);
+                output = tasks.findMechanism(keywordToFind);
+                break;
+            case 7:
+                //tell ui to parse and return task to make
+                //pass to tasklist to add the task only, no other computation needed
+                Task taskToAdd = ui.analyseTask(echo);
+                output = tasks.taskMechanism(taskToAdd);
+                storage.updateFile(taskToAdd, 0, 0);
+                break;
+            default:
+                break;
+            }
+        } catch (DukeException e) {
+            output = e.getMessage();
+        } catch (DateTimeParseException e) {
+            output = "Make sure your datetime format is correct! ";
+        } catch (IndexOutOfBoundsException e) {
+            output = "There are only: " + tasks.getSize()
+                    + " task(s) in the list to delete.\n";
+        }
+
+        return output;
+    }
+
```
this can be placed in the parser for better oop imo

Ragnapop commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,81 @@
+        int[] results = new int[] {0, 0};
+
+
+        if (echo.toLowerCase().equals("bye")) {
+            results[0] = 1;
+        } else if (echo.toLowerCase().equals("list")) {
+            results[0] = 2;
+        } else if (echo.matches("mark -?[0-9]*")) {
+            results[0] = 3;
+            results[1] = parser.digOutInt(echo);
+        } else if (echo.matches("unmark -?[0-9]*")) {
+            results[0] = 4;
+            results[1] = parser.digOutInt(echo);
+        } else if (echo.matches("delete -?[0-9]*")) {
+            results[0] = 5;
+            results[1] = parser.digOutInt(echo);
+        } else if (echo.matches("find [a-zA-Z0-9]*")) {
+            results[0] = 6;
+        } else {
+            results[0] = 7;
+        }
+
+        return results;
```
interesting method!

Ragnapop commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,81 @@
+     * @return a Task to be added later on
+     * @throws DukeException when user input is invalid
+     */
+    public Task analyseTask(String echo) throws DukeException {
+        Task taskAdded = new Task("Error. Unable to retrieve Task.");
+        String[] results = parser.decryptInput(echo);
+
+        if (results[1] == null && results[2] == null) {
+            taskAdded = new Todo(results[0]);
+        } else if (results[2] == null) {
+            taskAdded = new Deadline(results[0], results[1]);
+        } else {
+            taskAdded = new Event(results[0], results[1], results[2]);
+        }
+
+        return taskAdded;
```
wouldnt adding tasks to the tasklist be a function in the tasklist itself?

Ragnapop commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,104 @@
+     * @return the description of the Task after being unmarked
+     * @throws DukeException when task is already unmarked
+     */
+    public String unMarkTask() throws DukeException {
```
unmarkTask

Ragnapop reviewed (commented):
good job!

ChuaZenKhoon committed:
Act on Peer Review

Removed magic number and changed method name

ChuaZenKhoon committed:
Minor adjustments

ChuaZenKhoon committed:
Add assertions

Important assumptions are not confirmed.

By confirming assumptions at specific parts
of the code, we are able to narrow down
bug locations during runtime.

Let's add assertions when we
* add or delete a task
* update the data file
* use our case statement
* parse a task to be added

ChuaZenKhoon committed:
Update Code Quality

Current code quality can be improved

Better code quality increases
readability for users

Let's
* remove magic numbers
* remove commented out unused code

ChuaZenKhoon committed:
Merge pull request #1 from ChuaZenKhoon/branch-A-Assertions

Add assertions

ChuaZenKhoon committed:
Merge branch 'master' into branch-A-CodeQuality

ChuaZenKhoon committed:
Merge pull request #2 from ChuaZenKhoon/branch-A-CodeQuality

Update Code Quality

ChuaZenKhoon committed:
Add Snooze function

Current task manager does not
allow for users to snooze tasks

This function will give more
flexibility to users

Other functions are also
adjusted to fit this new functionality

ChuaZenKhoon committed:
Minor adjustments

.jar file was not working

Executable makes user experience
better

Let's fix the path for gradle
to build the shadowjar

ChuaZenKhoon committed:
Minor adjustments

Let's:
Fix bugs for Snooze
Adjust GUI size
Set up UI.png for later

ChuaZenKhoon committed:
Add User Guide

Users are not able to understand application

Let's add a User Guide!

ChuaZenKhoon committed:
Minor Adjustment

Fix bug for anchor links in UG

ChuaZenKhoon committed:
Minor adjustment

Fix markdown bug for
section anchor link for mark task

ChuaZenKhoon committed:
Minor Adjustment

Fixed bug where data file is not created

ChuaZenKhoon committed:
Update long methods and SLAP

Based on feedback and review
